Add typescript type declaration

sz-piotr opened the pull request:
Closes #2 

This PR adds typescript type declarations for the library. The declaration file tries to follow the order of exports from `index.js`.

Additionally it tries to fit seamlessly with the current build / release configuration by automatically copying the declaration file to `/dist` after compilation.

I would love to get feedback on this and get this as close as possible to the true types.

bcarroll22 commented:
Thanks for working on this! Since I’m not very familiar with typescript, I’d like to get some feedback from @TAGraves before we merge. Really appreciate the contribution!

TAGraves commented on src/index.d.ts:
```diff
@@ -0,0 +1,280 @@
+type Bound<T> = T extends (arg: any, ...rest: infer U) => infer V ? (...args: U) => V : never
+type BoundQueries<T> = { [P in keyof T]: Bound<T[P]> }
+
+export type FiberRoot = Omit<ReactTestInstance, 'find' | 'findAllByProps' | 'findAllByType' | 'findByProps' | 'findByType' | 'instance'>
```
Can you talk through your thinking in naming this `FiberRoot`? I was thinking something along the lines of `NativeInstance` or `NativeElement` although neither of those are quite right, either.

sz-piotr replied:
I am open to changing this. The `FiberRoot` was chosen to make the types consistent with documentation https://www.native-testing-library.com/docs/api-queries#bya11yhint

bcarroll22 replied:
This brings up a good point, this change will likely require updating the docs also so that the types match.

TAGraves commented on src/index.d.ts:
```diff
@@ -0,0 +1,280 @@
+  timeout?: number
+  interval?: number
+}
+export declare function waitForElement<T>(callback: () => T, options?: WaitOptions): Promise<T>
```
`waitForElement` also take `container` as an option.

bcarroll22 replied:
This was actually removed, it no longer takes a container 👍 

Unrelated to this PR, but the reason for this was that in `dom-testing-library` the container is used as the root of a `MutationObserver` but we don't actually observe for mutation. It was an option in previous versions, but it didn't actually do anything. This is true for ^1.6.0.

TAGraves commented on src/index.d.ts:
```diff
@@ -0,0 +1,280 @@
+// ------------
+
+// TODO: options are pretty-format options
+export declare function prettyPrint(element: ReactTestRenderer | FiberRoot, maxLength?: number, options?: any): string
```
You can `import { OptionsReceived } from 'pretty-format'` to get the options.

Also, technically `element` can be anything. I would at least make it `ReactTestRenderer | FiberRoot | string

sz-piotr commented:
Great. I'll probably get around to continue work on it on Monday. Also, do you have any suggestions on what should be the type for the NativeEvent second parameter? I left it as any, because I didn't really know what it should be

bcarroll22 commented:
NativeEvent is flexible because of handlers like `onChangeText`. It can either take something like:

```
new NativeEvent('change', { validTargets: ['TextInput'], nativeEvent: { text: 'hello world' } })
```

or:

```
new NativeEvent('changeText', 'hello world')
```

`target` would be the result of a query, so I suppose it'd be what we'd call `FiberRoot` or `NativeElement`.

`new NativeEvent().event` is intended to closely resemble what you'd receive from a real `SyntheticEvent` in react-native. So something like:

```
{ 
  type: 'PressEvent', 
  nativeEvent: { text: 'hello world' } 
}
```

but it'll also includes any other properties you passed to the constructor on the second arg. So for example:

```
new NativeEvent('change', { nativeEvent: { text: 'hello world' }, foo: 'bar' })
```

would give you this on `event.event`:

```
{ 
  type: 'change', 
  nativeEvent: { text: 'hello world' }, 
  foo: 'bar' 
}
```

sz-piotr commented:
Hey! I rebased the PR so that it includes all the changes made over the weekend. I also changed a few things:
* Renamed `FiberRoot` to `NativeTestInstance`. The reasoning is `(react|native)-testing-library` -> `(React|Native)TestInstance`
* Added `OptionsReceived` from `pretty-format`
* Added event initialisers. Decided to go with `NativeSyntheticEvent` as the base. Hopefully this makes sense

bcarroll22 commented:
Thanks again for all of this work. I’m good with getting this in and iterating, like @tagraves mentioned. Your reasoning makes sense! 

Next steps for me, I think because of the size of this library it would be a lot more maintainable if these were split into files like in `dom-testing-library` (there’s a lot going on in this compared to `react-testing-library`). That’s really my only feedback right now though. Thanks again for the hard work!

bcarroll22 commented:
:tada: This PR is included in version 1.7.0 :tada:

The release is available on:
- `npm package (@latest dist-tag)`
- [GitHub release](https://github.com/bcarroll22/native-testing-library/releases/tag/v1.7.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
